analysis/priestdiscipline/src/modules/spells/Penance.js

```diff
@@ -7,33 +7,16 @@ import BoringSpellValue from 'parser/ui/BoringSpellValue';
 import Statistic from 'parser/ui/Statistic';
 import STATISTIC_ORDER from 'parser/ui/STATISTIC_ORDER';
 
-const PENANCE_MINIMUM_RECAST_TIME = 3500; // Minimum duration from one Penance to Another
-
 class Penance extends Analyzer {
-  get missedBolts() {
-    return [...this.eventGrouper].reduce((missedBolts, cast) => {
-      cast.length > 3 ? (this._boltCount = 6) : (this._boltCount = 3); // checking for Penitent one legendary
-      return missedBolts + (this._boltCount - cast.length);
-    }, 0);
-  }
-
-  get casts() {
-    return [...this.eventGrouper].length;
-  }
-
-  get currentBoltNumber() {
-    return [...this.eventGrouper].slice(-1)[0].length - 1; // -1 here for legacy code
-  }
-
-  _boltCount = 3;
-  hits = 0;
-  eventGrouper = new EventGrouper(PENANCE_MINIMUM_RECAST_TIME);
+  _expectedBoltCount = 0;
+  _boltCount = 0;
+  _missedBolts = 0;
+  _casts = 0;
 
   constructor(options) {
     super(options);
 
-    // Castigation Penance bolt count to 4 (from 3)
-    this._boltCount = this.selectedCombatant.hasTalent(SPELLS.CASTIGATION_TALENT.id) ? 4 : 3;
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(SPELLS.PENANCE_CAST), this.onCast);
     this.addEventListener(Events.damage.by(SELECTED_PLAYER), this.onDamage);
     this.addEventListener(Events.heal.by(SELECTED_PLAYER), this.onHeal);
```

Comment:
We can add `.spell(SPELLS.PENANCE_HEAL)` to this event listener to filter out non penance healing events